core/src/client/mod.rs

```diff
@@ -592,6 +655,61 @@ impl<'a, R> IntoIterator for BatchResponse<'a, R> {
 	}
 }
 
+#[derive(thiserror::Error, Debug)]
+enum TrySubscriptionSendError {
+	#[error("The subscription is closed")]
+	Closed,
+	#[error("A subscription message was dropped")]
+	TooSlow(JsonValue),
+}
+
+#[derive(Debug)]
+pub(crate) struct SubscriptionSender {
+	inner: mpsc::Sender<JsonValue>,
+	lagged: SubscriptionLagged,
+}
+
+impl SubscriptionSender {
+	fn send(&self, msg: JsonValue) -> Result<(), TrySubscriptionSendError> {
+		match self.inner.try_send(msg) {
+			Ok(_) => Ok(()),
+			Err(TrySendError::Closed(_)) => Err(TrySubscriptionSendError::Closed),
+			Err(TrySendError::Full(m)) => {
+				self.lagged.set_missed();
+				Err(TrySubscriptionSendError::TooSlow(m))
+			}
+		}
+	}
+}
+
+#[derive(Debug)]
+pub(crate) struct SubscriptionReceiver {
+	inner: mpsc::Receiver<JsonValue>,
+	lagged: SubscriptionLagged,
```

Comment:
It's a shame we need a mutex here, because that's potentially polled a lot, so I wonder what perf impact it has.

Another way to implement this could be to have a channel with CAPACITY + 1, and then if we get up to CAPACITY then push the "Err(MissedMessages)" error to the final vacant slot. Then all subsequent messages are thrown away until the channel starts draining and every time it gets up to CAPACITY again we push another err on. But this could lead to silly things like a whole channel full of such errs if the client is too slow.

I still wonder whether we shouldn't just disconnect them when they are too slow. If they want to avoid it then they use something like in your example here https://github.com/paritytech/jsonrpsee/pull/1320/files#diff-0255f26eec4d19e28d74dcf6bd4ace6bfa5e4e45a1b1e9a63594548150227b60R68. That keeps the impl simple and avoids a mutex at least :)